packages/concerto-core/lib/decoratormanager.js

```diff
@@ -253,6 +253,178 @@ class DecoratorManager {
+     * @param {boolean} [options.validate] - validate that decorator command set is valid
+     * with respect to to decorator command set model
+     * @param {boolean} [options.validateCommands] - validate the decorator command set targets. Note that
+     * the validate option must also be true
+     * @param {boolean} [options.migrate] - migrate the decoratorCommandSet $class to match the dcs model version
+     * @returns {ModelManager} a new model manager with the decorations applied
+     */
+    static optimizedDecorateModels(modelManager, decoratorCommandSet, options) {
+
+        if (options?.migrate && this.canMigrate(decoratorCommandSet, DCS_VERSION)) {
+            decoratorCommandSet = this.migrateTo(decoratorCommandSet, DCS_VERSION);
+        }
+
+        if (options?.validate) {
+            const validationModelManager = new ModelManager({
+                strict: true,
+                metamodelValidation: true,
+                addMetamodel: true,
+            });
+            validationModelManager.addModelFiles(modelManager.getModelFiles());
+            validationModelManager.addCTOModel(
+                DCS_MODEL,
+                'decoratorcommands@0.3.0.cto'
+            );
+            const factory = new Factory(validationModelManager);
+            const serializer = new Serializer(factory, validationModelManager);
+            serializer.fromJSON(decoratorCommandSet);
+            if (options?.validateCommands) {
+                decoratorCommandSet.commands.forEach((command) => {
+                    DecoratorManager.validateCommand(
+                        validationModelManager,
+                        command
+                    );
+                });
+            }
+        }
+        const { namespaceCommandsMap, declarationCommandsMap, propertyCommandsMap, mapElementCommandsMap, typeCommandsMap }  = this.getDecoratorMaps(decoratorCommandSet);
+        const ast = modelManager.getAst(true);
+        const decoratedAst = JSON.parse(JSON.stringify(ast));
+        decoratedAst.models.forEach((model) => {
+            model.declarations.forEach((decl) => {
+                const { name: declarationName, $class: $classForDeclaration } = decl;
+                if (declarationCommandsMap.has(declarationName)) {
+                    const commands = declarationCommandsMap.get(declarationName);
+                    console.log('decl');
+                    commands.forEach(command => {
+                        this.executeDeclarationCommand(model.namespace, decl, command);
+                    });
+                }
+
+                if (namespaceCommandsMap.has(model.namespace)) {
+                    const commands = namespaceCommandsMap.get(model.namespace);
+                    console.log('name');
+                    commands.forEach(command => {
+                        this.executeDeclarationCommand(model.namespace, decl, command);
+                    });
+                }
+
```

Comment:
This should be outside the declarations.forEach loop

Reply:
https://github.com/accordproject/concerto/issues/858#issuecomment-2157341398